File: Trabalho 3/cpp/sorting.cpp
#include <iostream>
#include <fstream>
#include <time.h>
using namespace std;

typedef struct
{
    int *dado;
    unsigned n;
} TVetDin;

void initVetDin(TVetDin *V, unsigned size)
{
    V->dado = (int *)malloc(size * sizeof(int));
    if (!V->dado)
    {
        printf("Erro: Memória insuficiente!\n");
        exit(1);
    }
    else
    {
        V->n = size;
        return;
    }
}

void deleteVetDin(TVetDin *V)
{
    free(V->dado);
}

void printVetDin(TVetDin V)
{
    printf("[");
    for (unsigned i = 0; i < V.n; i++)
    {
        printf("%d", V.dado[i]);
        if (i < V.n - 1)
        {
            printf(", ");
        }
    }
    printf("]");
    return;
}

void insertionChange(TVetDin *V, int i)
{
    for (i; i > 0; i--)
    {
        if (V->dado[i] > V->dado[i - 1])
        {
            return;
        }
        int aux = V->dado[i];
        V->dado[i] = V->dado[i - 1];
        V->dado[i - 1] = aux;
    }
}

void insertionSort(TVetDin *V)
{
    for (int i = 0; i < V->n; i++)
    {
        insertionChange(V, i);
    }
}

int main()
{
    TVetDin lista;
    int n, x;
    float tempo;
    FILE *in;
    clock_t timeStart, timeEnd;

    in = fopen("../numeros.txt", "r");

    if (!in)
    {
        cout << "Arquivo não encontrado\n";
        exit(1);
    }

    cout << "Digite a quantidade de elementos" << endl;
    scanf("%d", &n);
    initVetDin(&lista, n);

    for (int i = 0; i < n; i++)
    {
        fscanf(in, "%d", &x);
        lista.dado[i];
    }

    timeStart = clock();

    timeEnd = clock();
    tempo = (float)(((timeEnd - timeStart) + 0.0) / CLOCKS_PER_SEC);

    printVetDin(lista);
    insertionSort(&lista);
    printVetDin(lista);

    printf("\n\nTempo:%.3f\n", tempo);
}

File: Trabalho 3/go/sorting.go
package main

import (
	"fmt"
	"os"
	"time"
)

type Node struct {
	data        int
	left, right *Node
}

type List struct {
	start, end *Node
	n          int
}

func (lista *List) insert(x int) {
	newNode := new(Node)

	if newNode == nil {
		panic("Não há memória suficiente para alocar um novo nó!")
	}

	newNode.data = x
	newNode.right, newNode.left = nil, nil

	if lista.n == 0 {
		lista.start = newNode
	} else {
		newNode.left = lista.end
		lista.end.right = newNode
	}

	lista.end = newNode
	lista.n++
}

func (lista *List) printList() {
	aux := lista.start
	fmt.Print("{")
	for aux != nil {
		fmt.Printf("%v", aux.data)
		if aux.right != nil {
			fmt.Print(", ")
		}
		aux = aux.right
	}
	fmt.Println("}")
}

func (lista *List) insertionSortE() {
	troca := lista.start
	for troca != nil {
		aux := troca.right
		lista.insertionChangeE(troca)
		troca = aux
	}
}

func (lista *List) insertionSort() {
	troca := lista.start
	for troca != nil {
		aux := troca.right
		lista.insertionChange(troca)
		troca = aux
	}
}

func (lista *List) insertionChange(change *Node) {
	if change.left == nil {
		return
	}
	aux := change.left
	for aux != nil {
		if change.data > aux.data {
			if change.right == nil {
				lista.end = change.left
			} else {
				change.right.left = change.left
			}

			change.left.right = change.right
			aux.right.left = change
			change.right = aux.right
			aux.right = change
			change.left = aux
			return
		}
		aux = aux.left
	}
	if change.right == nil {
		lista.end = change.left
	} else {
		change.right.left = change.left
	}
	change.left.right = change.right
	change.right = lista.start
	change.left = nil
	lista.start.left = change
	lista.start = change
}

func (lista *List) insertionChangeE(change *Node) {
	for change != nil {
		if change.left == nil {
			return
		}
		if change.data < change.left.data {
			change.data, change.left.data = change.left.data, change.data
			change = change.left
		} else {
			return
		}
	}
}

func change() {

}

func main() {
	var (
		lista, lista2 List
		n, x          int
	)

	in1, inError1 := os.Open("../numeros.txt")

	if inError1 != nil {
		fmt.Println("Erro:", inError1)
		in1.Close()
		return
	}

	in2, inError2 := os.Open("../numeros.txt")

	if inError2 != nil {
		fmt.Println("Erro:", inError2)
		in2.Close()
		return
	}

	fmt.Println("Insira a quantidade de elementos da lista: ")
	fmt.Scan(&n)

	for i := 0; i < n; i++ {
		fmt.Fscanf(in1, "%d", &x)
		lista.insert(x)
	}
	in1.Close()

	for i := 0; i < n; i++ {
		fmt.Fscanf(in2, "%d", &x)
		lista2.insert(x)
	}
	in2.Close()

	start1 := time.Now()

	// lista.printList()

	lista.insertionSortE()

	// fmt.Printf("\n\n\n\n")

	// lista.printList()

	duration1 := time.Since(start1)

	start2 := time.Now()
	lista2.insertionSort()
	duration2 := time.Since(start2)

	fmt.Printf("Duração de tempo 1 (troca um a um):")
	fmt.Println(duration1)

	fmt.Printf("Duração de tempo 2 (insiro na posição correta):")
	fmt.Println(duration2)

}
